omod/src/main/java/org/openmrs/module/pihmalawi/web/taglibs/ETraceAccessTag.java

```diff
@@ -0,0 +1,325 @@
+                    + "&inPopup=true'); return false;\">";
+        }
+        return link + "View " + f.getName() + "</a><br/>"
+                + (includeAppointmentInfo ? getDetails(p, initialEncounter) + "<br/>" : "")
+                + "(" + additionalMessage + ")";
+    }
+
+    protected String createNewCardHtmlTag(Patient p, Form f) {
+        String link = "";
+        String newMasterCardConfig = getNewMasterCardConfiguration(f);
+        if (newMasterCardConfig != null) {
+
+            String uuid = UUID.randomUUID().toString();
+
+            // special case here to allow us to pre-pick the encounter date
+            link = "<a onclick=\"window.location.href='"
+                    + newMasterCardConfig
+                    + "&patientId=" + p.getPatientId()
+                    + "&encounterDate=' + $j.datepicker.formatDate('yy-mm-dd', $j.datepicker.parseDate('dd/mm/yy', $j('#date-" + uuid + "').val()))"
+                    + "\">";
+
+            return link + "Create new " + f.getName() + "</a> on " + dateTag("date-" + uuid, "date-" + uuid);
+        }
+        else {
+
+            link = "<a href=\"/openmrs/module/htmlformentry/htmlFormEntry.form?personId="
+                    + p.getPersonId()
+                    + "&patientId="
+                    + p.getPatientId()
+                    + "&returnUrl=%2fopenmrs%2fpatientDashboard.form&formId="
+                    + f.getFormId() + "\">";
+
+            return link + "Create new " + f.getName() + "</a>";
+        }
+    }
+
+    protected String createEditCardHtmlTag(Patient p, Form f, Encounter initialEncounter) {
+        String link = "";
+        String newMasterCardConfig = getNewMasterCardConfiguration(f);
+        if (newMasterCardConfig != null) {
+            link = "<a href=\"" + newMasterCardConfig + "&patientId="+p.getPatientId()+"\">";
+        }
+        else {
+            link = "<a href=\"/openmrs/module/htmlformentry/htmlFormEntry.form?encounterId=" + initialEncounter.getId() + "&mode=EDIT\">";
+        }
+
+        return link + "Edit " + f.getName() + "</a><br/>" + (includeAppointmentInfo ? getDetails(p, initialEncounter) + "<br/>" : "");
+
+    }
+
+
+
+    protected String getDetails(Patient p, Encounter initialEncounter) {
+        EncounterType followupEncounterType = null;
+        if (StringUtils.isNotBlank(getFollowupEncounterTypeName())) {
+            followupEncounterType = Context.getEncounterService().getEncounterType(getFollowupEncounterTypeName());
+        }
+        if (followupEncounterType == null) {
+            followupEncounterType = Context.getEncounterService().getEncounterType(getFollowupEncounterTypeId());
+        }
+        List<Encounter> followups = Context.getEncounterService().getEncounters(p, null, null, null, null, Arrays.asList(followupEncounterType), null, false);
+        String created = "Created: " + Helper.formatDate(initialEncounter.getEncounterDatetime());
+        String visited = "Visited: no";
```

Comment:
This verbage is wrong, this should be re-worded to something like "Last Tracking Date"